### DISCUSSION TITLE
How to asynchronously retrieve relationship-related data (many_to_many or many_to_one)

### DISCUSSION THREAD

**Discussion by t-kaper** [sqlalchemy/sqlalchemy#8005] [Usage Questions]
Hi there, how do I work with association_table asynchronously?  It used to work this way for me.

MODEL

```python

association_user_role = Table(
    'association_user_role',
    Base.metadata,
    Column('user_id', ForeignKey('user.id'), primary_key=True),
    Column('user_role_id', ForeignKey('user_role.id'), primary_key=True)
)


class User(Base):
    __tablename__ = "user"  # type: ignore
    __mapper_args__ = {"eager_defaults": True}
    id = Column(Integer, primary_key=True, autoincrement=True)
    email = Column(String, unique=True)
    password = Column(String)
    user_role = relationship("UserRole", secondary=association_user_role)

    def __repr__(self):
        return f"User (email={self.email},username={self.username})"


class UserRole(Base):
    __tablename__ = "user_role"  # type: ignore
    __mapper_args__ = {"eager_defaults": True}

    id = Column(Integer, primary_key=True)
    role = Column(String, unique=True)

    def __repr__(self):
        return f"User Role(id={self.id},role={self.role})"

```
 
MY CODE

```python

user_role = await UserRole.get(db, id=USER_ROLE['user'])
user1 = await User.get_user_by_email(db, _email)
await user1.user_role.add(user_role)
```

But the line await user1.user_role.add(user_role)
doesn't work and completes with an error. I found that there is a way to work with relationship. 

```python
 query = db.query(self.model).options(
            selectinload(User.user_role))
```

But I don't quite understand how to save data correctly.
there are good examples of how to write code correctly. 

 Will this situation persist?  It's just that it used to be more nativethey were

**Comment by CaselIT** [comment]
> ```python
> user_role = await UserRole.get(db, id=USER_ROLE['user'])
> user1 = await User.get_user_by_email(db, _email)
> await user1.user_role.add(user_role)
> ```

I don't know that this code does, it's not using the sqlalchemy api, so I don't know how to help.

I would suggest reading the documentation on asyncio: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html
